scipy/stats/stats.py

```diff
@@ -4153,6 +4154,177 @@ def binary_search(n, n1, n2, side):
+        alternative='two-sided',
+        POOLED=True,
+        num_it=4):
+    """
+    Perform a Barnard exact test on a 2x2 contingency table.
+
+    Parameters
+    ----------
+    table : array_like of ints
+        A 2x2 contingency table.  Elements should be non-negative integers.
+    alternative : {'two-sided', 'less', 'greater'}, optional
+        Defines the alternative hypothesis.
+        The following options are available (default is 'two-sided'):
+
+          * 'two-sided'
+          * 'less': one-sided
+          * 'greater': one-sided
+
+    POOLED : bool, optional
+        Whether to compute score statistic with pooled variance or
+        unpooled (Wald statistic)
+
+    nuisance_num : int, optional
+        Number of nuisance's samples to generate. Default is 1000. Must be
+        non-negative
+
+    Returns
+    -------
+    score_statistic : float
+        Z statistic with pooled (Score) or unpooled (Wald) variance, depending
+        on the user choice of `POOLED` param.
+    p_value : float
+        P-value, the probability of obtaining a distribution at least as
+        extreme as the one that was actually observed, assuming that the
+        null hypothesis is true.
+
+    See Also
+    --------
+    chi2_contingency : Chi-square test of independence of variables in a
+        contingency table.
+    fisher_exact : Fisher exact test on a 2x2 contingency table.
+
+    References
+    ----------
+    .. [1] `Mehta, Cyrus & Senchaudhuri, Pralay. (2003). Conditional
+    versus Unconditional Exact Tests for Comparing Two Binomials`__.
+
+    .. __: https://www.statsols.com/hubfs/Resources_/Comparing-Two-Binomials.pdf
+
+    Examples
+    --------
+    examples from [1]_.
+
+    Consider the following example of a vaccine efficacy study (Chan, 1998).
+    In a randomized clinical trial of 30 subjects, 15 were innoculated with a
+    recombinant DNA influenza vaccine and the 15 were innoculated with a
+    placebo. Twelve of the 15 subjects in the placebo group (80%) eventually
+    became infected with influenza whereas for the vaccine group, only 7 of
+    the 15 subjects (47%) became infected. The data are tabulated as a 2 × 2
+    table::
+
+                    Vaccine  Placebo
+        Yes             8        2
+        No              1        5
+
+    We use this table to find the p-value:
+    >>> import scipy.stats as stats
+    >>> TX, pvalue = stats.barnard_exact([[7, 12], [8, 3]])
+    >>> pvalue
+    0.0681...
+
+    """
+    nuisance_num = 100
+
+    if num_it < 0:
+        raise ValueError(
+            "Number of iterations must be non-negative, "
+            f"found {num_it!r}"
+        )
+
+    c = np.asarray(
+        table,
+        dtype=np.int64)
+
+    if not c.shape == (2, 2):
+        raise ValueError("The input `table` must be of shape (2, 2).")
+
+    if np.any(c < 0):
+        raise ValueError("All values in `table` must be nonnegative.")
+
+    if 0 in c.sum(axis=0) or 0 in c.sum(axis=1):
```

Comment:
In Fisher test row sums and col sums have to be positive. The barnard test is defined when on of rows sums is null. Therefore only column sums should be checked.